File: kanonak-wire/java/conformance/Conformance.java
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.kanonak.wire.WireError;
import org.kanonak.wire.WireReader;
import org.kanonak.wire.WireWriter;

/**
 * Drives the shared wire vectors through the Java kanonak-wire port. Read
 * vectors run an op-script against a hex buffer asserting values or required
 * errors {kind, offset}; write vectors run writer ops asserting exact output
 * bytes. Cases requiring a capability Java lacks are skipped WITH a reported
 * skip count.
 */
public final class Conformance {
    // Java has wide numeric params (int/long can exceed the wire range) and
    // UTF-16 strings (a String can hold an unpaired surrogate); it lacks
    // dynamic-numeric (2.5 is not representable in int).
    static final Set<String> CAPABILITIES = Set.of("wide-numeric-params", "utf16-strings");

    static int pass = 0;
    static int fail = 0;
    static int skipped = 0;

    public static void main(String[] args) throws Exception {
        String vdir = args.length > 0 ? args[0] : "../vectors";
        int fails = run(Paths.get(vdir, "wire-vectors.json"));
        System.exit(fails == 0 ? 0 : 1);
    }

    @SuppressWarnings("unchecked")
    static int run(Path path) throws Exception {
        Map<String, Object> doc = (Map<String, Object>) Json.parse(Files.readString(path, StandardCharsets.UTF_8));
        List<Object> readVectors = (List<Object>) doc.get("readVectors");
        List<Object> writeVectors = (List<Object>) doc.get("writeVectors");

        for (Object o : readVectors) {
            runReadVector((Map<String, Object>) o);
        }
        for (Object o : writeVectors) {
            runWriteVector((Map<String, Object>) o);
        }

        int total = readVectors.size() + writeVectors.size();
        System.out.println("wire-vectors: " + pass + "/" + total + " pass (" + skipped + " skipped)");
        if (fail > 0) {
            System.err.println(fail + " VECTOR(S) FAILED");
            return fail;
        }
        System.out.println("ALL VECTORS PASS");
        return 0;
    }

    static void failCase(String id, String msg) {
        fail++;
        System.err.println(id + ": " + msg);
    }

    @SuppressWarnings("unchecked")
    static boolean checkError(String id, String opName, RuntimeException e, Map<String, Object> want) {
        if (!(e instanceof WireError we)) {
            failCase(id, opName + ": threw a non-WireError: " + e.getMessage());
            return false;
        }
        String wantKind = (String) want.get("kind");
        if (!we.kind.equals(wantKind)) {
            failCase(id, opName + ": expected error kind " + wantKind + ", got " + we.kind + " (" + we.getMessage() + ")");
            return false;
        }
        Object wantOffset = want.get("offset");
        if (wantOffset != null) {
            Integer wanted = ((Number) wantOffset).intValue();
            if (!wanted.equals(we.offset)) {
                failCase(id, opName + ": expected error offset " + wanted + ", got " + we.offset);
                return false;
            }
        }
        return true;
    }

    // -- Read vectors -----------------------------------------------------

    /** Returns Long for numeric reads, String for byte/text reads, null for void ops. */
    static Object runReadOp(WireReader r, Map<String, Object> op) {
        String name = (String) op.get("op");
        switch (name) {
            case "u8": return (long) r.u8();
            case "u16be": return (long) r.u16be();
            case "u32be": return r.u32be();
            case "bytes": return hex(r.bytes(((Number) op.get("n")).intValue()));
            case "uuid": return r.uuid();
            case "utf8": return r.utf8(((Number) op.get("n")).intValue());
            case "lenPrefixedBytes16": return hex(r.lenPrefixedBytes16());
            case "rest": return hex(r.rest());
            case "remaining": return (long) r.remaining();
            case "expectEnd": r.expectEnd(); return null;
            default: throw new IllegalStateException("conformance: unknown read op '" + name + "'");
        }
    }

    @SuppressWarnings("unchecked")
    static void runReadVector(Map<String, Object> v) {
        String requires = (String) v.get("requires");
        if (requires != null && !CAPABILITIES.contains(requires)) {
            skipped++;
            return;
        }
        String id = (String) v.get("id");
        WireReader r = new WireReader(hexToBytes((String) v.get("bytes")));
        boolean ok = true;
        for (Object o : (List<Object>) v.get("ops")) {
            Map<String, Object> op = (Map<String, Object>) o;
            String opName = (String) op.get("op");
            Map<String, Object> expectError = (Map<String, Object>) op.get("expectError");
            if (expectError != null) {
                try {
                    runReadOp(r, op);
                    failCase(id, opName + ": expected " + expectError.get("kind") + ", got a value");
                    ok = false;
                } catch (RuntimeException e) {
                    if (!checkError(id, opName, e, expectError)) ok = false;
                }
                break; // an error op ends the script
            }
            Object got;
            try {
                got = runReadOp(r, op);
            } catch (RuntimeException e) {
                failCase(id, opName + ": threw " + e.getMessage());
                ok = false;
                break;
            }
            Object expected = op.get("expected");
            if (expected != null && !valueMatches(expected, got)) {
                failCase(id, opName + ": expected " + expected + ", got " + got);
                ok = false;
                break;
            }
        }
        if (ok) pass++;
    }

    static boolean valueMatches(Object expected, Object got) {
        if (expected instanceof Number n) {
            return got instanceof Long l && n.longValue() == l;
        }
        return expected.equals(got);
    }

    // -- Write vectors ----------------------------------------------------

    static void runWriteOp(WireWriter w, Map<String, Object> op) {
        String name = (String) op.get("op");
        switch (name) {
            case "u8": w.u8(((Number) op.get("value")).intValue()); return;
            case "u16be": w.u16be(((Number) op.get("value")).intValue()); return;
            case "u32be": w.u32be(((Number) op.get("value")).longValue()); return;
            case "bytes": w.bytes(hexToBytes((String) op.get("hex"))); return;
            case "uuid": w.uuid((String) op.get("value")); return;
            case "utf8": w.utf8(utf8OpString(op)); return;
            case "lenPrefixedBytes16": w.lenPrefixedBytes16(hexToBytes((String) op.get("hex"))); return;
            default: throw new IllegalStateException("conformance: unknown write op '" + name + "'");
        }
    }

    @SuppressWarnings("unchecked")
    static String utf8OpString(Map<String, Object> op) {
        List<Object> codeUnits = (List<Object>) op.get("utf16CodeUnits");
        if (codeUnits == null) {
            return (String) op.get("value");
        }
        char[] chars = new char[codeUnits.size()];
        for (int i = 0; i < chars.length; i++) {
            chars[i] = (char) ((Number) codeUnits.get(i)).intValue();
        }
        return new String(chars);
    }

    @SuppressWarnings("unchecked")
    static void runWriteVector(Map<String, Object> v) {
        String requires = (String) v.get("requires");
        if (requires != null && !CAPABILITIES.contains(requires)) {
            skipped++;
            return;
        }
        String id = (String) v.get("id");
        WireWriter w = new WireWriter();
        boolean ok = true;
        for (Object o : (List<Object>) v.get("ops")) {
            Map<String, Object> op = (Map<String, Object>) o;
            String opName = (String) op.get("op");
            Map<String, Object> expectError = (Map<String, Object>) op.get("expectError");
            if (expectError != null) {
                try {
                    runWriteOp(w, op);
                    failCase(id, opName + ": expected " + expectError.get("kind") + ", got success");
                    ok = false;
                } catch (RuntimeException e) {
                    if (!checkError(id, opName, e, expectError)) ok = false;
                }
                break;
            }
            try {
                runWriteOp(w, op);
            } catch (RuntimeException e) {
                failCase(id, opName + ": threw " + e.getMessage());
                ok = false;
                break;
            }
        }
        String expectedBytes = (String) v.get("expectedBytes");
        if (ok && expectedBytes != null) {
            String got = hex(w.toBytes());
            if (!got.equals(expectedBytes)) {
                failCase(id, "expected bytes " + expectedBytes + ", got " + got);
                ok = false;
            }
        }
        if (ok) pass++;
    }

    // -- Hex helpers ------------------------------------------------------

    static byte[] hexToBytes(String hex) {
        byte[] out = new byte[hex.length() / 2];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }

    static String hex(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) {
            sb.append(String.format("%02x", x & 0xFF));
        }
        return sb.toString();
    }

    static String hex(ByteBuffer b) {
        StringBuilder sb = new StringBuilder(b.remaining() * 2);
        for (int i = b.position(); i < b.limit(); i++) {
            sb.append(String.format("%02x", b.get(i) & 0xFF));
        }
        return sb.toString();
    }

    // -- Minimal JSON parser (objects/arrays/strings/numbers/booleans/null) ----

    static final class Json {
        private final String s;
        private int i;

        private Json(String s) { this.s = s; }

        static Object parse(String s) {
            Json j = new Json(s);
            j.ws();
            Object v = j.value();
            j.ws();
            return v;
        }

        private Object value() {
            char c = s.charAt(i);
            switch (c) {
                case '{': return object();
                case '[': return array();
                case '"': return string();
                case 't': i += 4; return Boolean.TRUE;
                case 'f': i += 5; return Boolean.FALSE;
                case 'n': i += 4; return null;
                default: return number();
            }
        }

        private Map<String, Object> object() {
            Map<String, Object> m = new java.util.LinkedHashMap<>();
            i++; ws();
            if (s.charAt(i) == '}') { i++; return m; }
            while (true) {
                ws();
                String key = string();
                ws(); i++; // ':'
                ws();
                m.put(key, value());
                ws();
                char c = s.charAt(i++);
                if (c == '}') return m;
            }
        }

        private List<Object> array() {
            List<Object> a = new ArrayList<>();
            i++; ws();
            if (s.charAt(i) == ']') { i++; return a; }
            while (true) {
                ws();
                a.add(value());
                ws();
                char c = s.charAt(i++);
                if (c == ']') return a;
            }
        }

        private String string() {
            i++; // opening quote
            StringBuilder sb = new StringBuilder();
            while (true) {
                char c = s.charAt(i++);
                if (c == '"') return sb.toString();
                if (c == '\\') {
                    char e = s.charAt(i++);
                    switch (e) {
                        case '"': sb.append('"'); break;
                        case '\\': sb.append('\\'); break;
                        case '/': sb.append('/'); break;
                        case 'b': sb.append('\b'); break;
                        case 'f': sb.append('\f'); break;
                        case 'n': sb.append('\n'); break;
                        case 'r': sb.append('\r'); break;
                        case 't': sb.append('\t'); break;
                        case 'u':
                            sb.append((char) Integer.parseInt(s.substring(i, i + 4), 16));
                            i += 4;
                            break;
                        default: sb.append(e);
                    }
                } else {
                    sb.append(c);
                }
            }
        }

        private Double number() {
            int start = i;
            while (i < s.length() && "+-0123456789.eE".indexOf(s.charAt(i)) >= 0) i++;
            return Double.parseDouble(s.substring(start, i));
        }

        private void ws() {
            while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
        }
    }
}

File: kanonak-wire/rust/tests/conformance.rs
//! Drives the shared wire vectors through the Rust kanonak-wire port.

use kanonak_wire::*;
use serde_json::Value as J;
use std::fs;
use std::path::PathBuf;

/// Rust writer params are exact-width types and `&str` is always well-formed
/// UTF-8: none of the representability capabilities apply.
const CAPABILITIES: &[&str] = &[];

fn vectors_file() -> PathBuf {
    // tests run from the crate root (rust/); vectors are at ../vectors.
    let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    p.push("..");
    p.push("vectors");
    p.push("wire-vectors.json");
    p
}

fn hex_to_bytes(hex: &str) -> Vec<u8> {
    (0..hex.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).unwrap())
        .collect()
}

fn bytes_to_hex(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

fn check_error(id: &str, op: &str, e: &WireError, want: &J, fails: &mut usize) {
    let want_kind = want.get("kind").and_then(|k| k.as_str()).unwrap_or("");
    if e.kind.as_str() != want_kind {
        eprintln!("{id}: {op}: expected error kind {want_kind}, got {} ({e})", e.kind.as_str());
        *fails += 1;
        return;
    }
    if let Some(want_offset) = want.get("offset").and_then(|o| o.as_u64()) {
        if e.offset != Some(want_offset as usize) {
            eprintln!(
                "{id}: {op}: expected error offset {want_offset}, got {:?}",
                e.offset
            );
            *fails += 1;
        }
    }
}

enum ReadResult {
    Num(u64),
    Str(String),
    Hex(String),
    Unit,
}

fn run_read_op(r: &mut WireReader, op: &J) -> Result<ReadResult, WireError> {
    let name = op.get("op").and_then(|o| o.as_str()).unwrap_or("");
    let n = op.get("n").and_then(|n| n.as_u64()).unwrap_or(0) as usize;
    match name {
        "u8" => Ok(ReadResult::Num(r.u8()? as u64)),
        "u16be" => Ok(ReadResult::Num(r.u16_be()? as u64)),
        "u32be" => Ok(ReadResult::Num(r.u32_be()? as u64)),
        "bytes" => Ok(ReadResult::Hex(bytes_to_hex(r.bytes(n)?))),
        "uuid" => Ok(ReadResult::Str(r.uuid()?)),
        "utf8" => Ok(ReadResult::Str(r.utf8(n)?.to_string())),
        "lenPrefixedBytes16" => Ok(ReadResult::Hex(bytes_to_hex(r.len_prefixed_bytes16()?))),
        "rest" => Ok(ReadResult::Hex(bytes_to_hex(r.rest()))),
        "remaining" => Ok(ReadResult::Num(r.remaining() as u64)),
        "expectEnd" => {
            r.expect_end()?;
            Ok(ReadResult::Unit)
        }
        other => panic!("conformance: unknown read op '{other}'"),
    }
}

fn expected_matches(id: &str, opname: &str, got: &ReadResult, expected: &J, fails: &mut usize) {
    let ok = match got {
        ReadResult::Num(v) => expected.as_u64() == Some(*v),
        ReadResult::Str(s) | ReadResult::Hex(s) => expected.as_str() == Some(s.as_str()),
        ReadResult::Unit => true,
    };
    if !ok {
        let got_str = match got {
            ReadResult::Num(v) => v.to_string(),
            ReadResult::Str(s) | ReadResult::Hex(s) => s.clone(),
            ReadResult::Unit => "()".into(),
        };
        eprintln!("{id}: {opname}: expected {expected}, got {got_str}");
        *fails += 1;
    }
}

fn run_write_op(w: &mut WireWriter, op: &J) -> Result<(), WireError> {
    let name = op.get("op").and_then(|o| o.as_str()).unwrap_or("");
    match name {
        "u8" => {
            w.u8(op["value"].as_u64().unwrap() as u8);
            Ok(())
        }
        "u16be" => {
            w.u16_be(op["value"].as_u64().unwrap() as u16);
            Ok(())
        }
        "u32be" => {
            w.u32_be(op["value"].as_u64().unwrap() as u32);
            Ok(())
        }
        "bytes" => {
            w.bytes(&hex_to_bytes(op["hex"].as_str().unwrap()));
            Ok(())
        }
        "uuid" => w.uuid(op["value"].as_str().unwrap()).map(|_| ()),
        "utf8" => {
            w.utf8(op["value"].as_str().unwrap());
            Ok(())
        }
        "lenPrefixedBytes16" => w
            .len_prefixed_bytes16(&hex_to_bytes(op["hex"].as_str().unwrap()))
            .map(|_| ()),
        other => panic!("conformance: unknown write op '{other}'"),
    }
}

#[test]
fn wire_vectors() {
    let data: J = serde_json::from_str(&fs::read_to_string(vectors_file()).unwrap()).unwrap();
    let read_vectors = data["readVectors"].as_array().unwrap();
    let write_vectors = data["writeVectors"].as_array().unwrap();

    let mut pass = 0usize;
    let mut fails = 0usize;
    let mut skipped = 0usize;

    for v in read_vectors {
        let id = v["id"].as_str().unwrap();
        if let Some(req) = v.get("requires").and_then(|r| r.as_str()) {
            if !CAPABILITIES.contains(&req) {
                skipped += 1;
                continue;
            }
        }
        let bytes = hex_to_bytes(v["bytes"].as_str().unwrap());
        let mut r = WireReader::new(&bytes);
        let before = fails;
        for op in v["ops"].as_array().unwrap() {
            let opname = op["op"].as_str().unwrap();
            if let Some(want) = op.get("expectError") {
                match run_read_op(&mut r, op) {
                    Ok(_) => {
                        eprintln!("{id}: {opname}: expected {}, got a value", want["kind"]);
                        fails += 1;
                    }
                    Err(e) => check_error(id, opname, &e, want, &mut fails),
                }
                break;
            }
            match run_read_op(&mut r, op) {
                Ok(got) => {
                    if let Some(expected) = op.get("expected") {
                        expected_matches(id, opname, &got, expected, &mut fails);
                        if fails > before {
                            break;
                        }
                    }
                }
                Err(e) => {
                    eprintln!("{id}: {opname}: threw {e}");
                    fails += 1;
                    break;
                }
            }
        }
        if fails == before {
            pass += 1;
        }
    }

    for v in write_vectors {
        let id = v["id"].as_str().unwrap();
        if let Some(req) = v.get("requires").and_then(|r| r.as_str()) {
            if !CAPABILITIES.contains(&req) {
                skipped += 1;
                continue;
            }
        }
        let mut w = WireWriter::new();
        let before = fails;
        for op in v["ops"].as_array().unwrap() {
            let opname = op["op"].as_str().unwrap();
            if let Some(want) = op.get("expectError") {
                match run_write_op(&mut w, op) {
                    Ok(()) => {
                        eprintln!("{id}: {opname}: expected {}, got success", want["kind"]);
                        fails += 1;
                    }
                    Err(e) => check_error(id, opname, &e, want, &mut fails),
                }
                break;
            }
            if let Err(e) = run_write_op(&mut w, op) {
                eprintln!("{id}: {opname}: threw {e}");
                fails += 1;
                break;
            }
        }
        if fails == before {
            if let Some(expected) = v.get("expectedBytes").and_then(|e| e.as_str()) {
                let got = bytes_to_hex(w.as_bytes());
                if got != expected {
                    eprintln!("{id}: expected bytes {expected}, got {got}");
                    fails += 1;
                }
            }
        }
        if fails == before {
            pass += 1;
        }
    }

    let total = read_vectors.len() + write_vectors.len();
    println!("wire-vectors: {pass}/{total} pass ({skipped} skipped)");
    assert_eq!(fails, 0, "{fails} VECTOR(S) FAILED");
}
